### ISSUE TITLE
twisted.conch.client.unix incorrectly uses os.path.expanduser

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3061]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3061|
|Type|defect|
|Created|2008-02-29 16:29:13Z|
|Branch|https://github.com/twisted/twisted/tree/conch-unix-home-3061-3|

Basically, it doesn't try to find the home of the given user, but use the home of the owner of the running process to create a unix socket.

To reproduce:

```
HOME=/foo trial twisted.conch.test.test_conch.UnixClientTestCase
```

<details><summary>Searchable metadata</summary>

```
trac-id__3061 3061
type__defect defect
reporter__therve therve
priority__high high
milestone__ 
branch__branches_conch_unix_home_3061_3 branches/conch-unix-home-3061-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1204302553000000 1204302553000000
changetime__1205779942000000 1205779942000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [22748]) Branching to 'conch-unix-home-3061'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

This is ready to review. In particular, it should fix the 3 ubuntu63 buildbot slaves.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

The changes are almost entirely in the test suite.  It seems like the implementation should be fixed so that if HOME is not writable (or does not exist, or is not even set) something sensible happens.  Otherwise, we're just changing the tests to not cover the case that's really broken.

Am I missing anything?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [22758]) Manage exceptions in listen, add a test for a non existent HOME.

Refs [#3061](https://github.com/twisted/twisted/issues/3061)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Replying to [exarkun](#note_3):
> The changes are almost entirely in the test suite.  It seems like the implementation should be fixed so that if HOME is not writable (or does not exist, or is not even set) something sensible happens.  Otherwise, we're just changing the tests to not cover the case that's really broken.

Well, in fact something sensible happened: it failed. It just not fail very clearly :)

> Am I missing anything?

I was just being lazy I guess. I added another case for checking the error, and rewrap the error a bit better.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @therve|
|-|-|

If everything is normal, it should conflict.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [22786]) Branching to 'conch-unix-home-3061-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bigdog set owner to bigdog<br>bigdog set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|bigdog commented|
|-|-|

code cleanup is nice

fixes scoped to ticket

test cases cover changes

verified exarkun's earlier review comments were addressed

tested on windows and unix with no ill side effects

ready to submit code

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bigdog set owner to @therve<br>bigdog set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [22797]) Merge conch-unix-home-3061-2

Author: therve
Reviewers: bigdog, exarkun
Fixes [#3061](https://github.com/twisted/twisted/issues/3061)

Fix handling of missing HOME in twisted.conch.client.unix. It also fixes the
tests so that they define a custom HOME if necessary, and redefines the path
of the unix socket if the default path is too long.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `reopened`|
|-|-|

(In [22798]) Revert r22797: regressions under OS X

Reopens [#3061](https://github.com/twisted/twisted/issues/3061)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [22800]) Branching to 'conch-unix-home-3061-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

The failures under OS X were due to weird rights in /tmp. But after looking around, I switched to relative path, and it seems to do the trick. Please review.

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bigdog set owner to bigdog<br>bigdog set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|bigdog set owner to @therve<br>bigdog set status to `new`|
|-|-|

Sorry, I missed the OSX,  I only ran on windows and linux.  This review will utilize all active buildbots.

Scope of this review: additional changes to branch 3, previous branches were reviewed.

1) The path length reduction to 70 follows the discussion on IRC

2) test doc string updated to reflect relative path approach

3) switch to relative is a nice approach to the root problem

4) buildbots launched on active build bots with no regressions, and OSX buildbots were green.   
 
one bot is still running, if this is green, you are ready to merge. http://twistedmatrix.com/buildbot/builders/osx-py2.4-select-gc/builds/602


good work


ready for merge

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

  * `SSHTestConnectionForUnix.stopDeferred` should be documented (particularly, when it fires and why)
  * `twisted/conch/test/test_cftp.py`, `CFTPClientTestBase` probably doesn't need `tearDown` at all, you could delete it if you want.

Otherwise, looks good.  Merge when you think it's ready.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set status to `closed`|
|-|-|

(In [22882]) Merge conch-unix-home-3061-3

Author: therve
Reviewers: bigdog, exarkun
Fixes [#3061](https://github.com/twisted/twisted/issues/3061)

Fix handling of missing HOME in twisted.conch.client.unix. It also fixes the
tests so that they define a custom HOME if necessary, and redefines the path
of the unix socket if the default path is too long.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
